### ISSUE TITLE
Darwin stdenv is broken on staging

### ISSUE THREAD

**Issue by alyssais** [NixOS/nixpkgs#154810]
### Describe the bug

[Every recent staging PR](https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+is%3Aopen+base%3Astaging) that has had OfBorg test Darwin builds has failed while bootstrapping.  This means that it's not possible to see whether PRs will break Darwin builds or not.  Errors I've seen are

#### x86_64-darwin

```
output '/nix/store/ynngj9vlnl6q5bd5v63yrhpvx4vwq48n-bootstrap-stage2-stdenv-darwin' is not allowed to refer to the following paths:
  /nix/store/5p5pncjap0kax6jzsh5jxq2nz5h75b7y-zstd-1.5.1
  /nix/store/m4vzwxx1qvvbpslisljzz35xx0n5ay3w-libunistring-0.9.10
  /nix/store/x7x5s166m6dn6vanr0zcwm9pkrprvsbg-libidn2-2.3.2
```

#### aarch64-darwin

```
tar: This does not look like a tar archive
tar: Exiting with failure status due to previous errors
do not know how to unpack source archive /nix/store/is8fadziysa55wpq1dppzhv5yhlhhm6n-coreutils-9.0.tar.xz
error: builder for '/nix/store/vzjacfky07pxd09w1dlz0abylmnv075z-coreutils-9.0.drv' failed with exit code 1;
```

### Notify maintainers

@NixOS/darwin-maintainers

**Comment by veprbl** [comment]
This is a regression from #145610

**Comment by Izorkin** [comment]
Build error due to zstd compression support in curl?

**Comment by reckenrode** [comment]
I did a `git bisect`. It reported 29526bc2eb9463515d0568c05e28b8a3bec09c91 as the first bad commit.

**Comment by veprbl** [comment]
I'm not sure what is the deal with the aarch64-darwin failure. Can somebody reproduce? We had some file corruption occurring on Hydra, I wonder if that's it again.

**Comment by uri-canva** [comment]
I get a completely different error on `aarch64-darwin` when building `curl` on https://github.com/NixOS/nixpkgs/commit/29526bc2eb9463515d0568c05e28b8a3bec09c91:
```
building '/nix/store/c5d6fgwfm7vq8qyggxgyf09j3nqxl1a7-python3-minimal-3.9.9.drv'...
unpacking sources
unpacking source archive /nix/store/ka6ph2kwfm5g75zxxvvg8y08zryyh9b7-Python-3.9.9.tar.xz
source root is Python-3.9.9
setting SOURCE_DATE_EPOCH to timestamp 1636999517 of file Python-3.9.9/Misc/NEWS
patching sources
applying patch /nix/store/p4qar37zn2892m83b2lkp8fw5dsjkzxy-no-ldconfig.patch
patching file Lib/ctypes/util.py
applying patch /nix/store/r112dk8w7zvdjipki58ch00m825li7fq-virtualenv-permissions.patch
patching file Lib/venv/__init__.py
Hunk #1 succeeded at 402 with fuzz 2 (offset 23 lines).
applying patch /nix/store/dkb2rjyj7lwmvsn4zzwx85kx8r61nk9w-darwin-libutil.patch
patching file Modules/posixmodule.c
Hunk #1 succeeded at 7148 (offset 1268 lines).
applying patch /nix/store/x6aiw4vay2b63slqz5byimn0alhg5b1s-darwin-tcl-tk.patch
patching file setup.py
Hunk #1 succeeded at 1988 (offset 7 lines).
applying patch /nix/store/xid1jwdpp3zvg7a9zl41snb9hgvam1w3-bpo-44689-ctypes.util.find_library-now-finds-macOS-1.patch
patching file Misc/NEWS.d/next/macOS/2021-07-20-22-27-01.bpo-44689.mmT_xH.rst
patching file Modules/_ctypes/callproc.c
Reversed (or previously applied) patch detected!  Assume -R? [n]
Apply anyway? [n]
Skipping patch.
2 out of 2 hunks ignored -- saving rejects to file Modules/_ctypes/callproc.c.rej
builder for '/nix/store/c5d6fgwfm7vq8qyggxgyf09j3nqxl1a7-python3-minimal-3.9.9.drv' failed with exit code 1
```

**Comment by veprbl** [comment]
Not surprising that the error is different, the aarch64-darwin doesn't use a pure CoreFoundation build, so its stdenv won't depend on `curl`:
https://github.com/NixOS/nixpkgs/blob/bc59ba15b64d0a0ee1d1764f18b4f3480d2c3e5a/pkgs/top-level/darwin-packages.nix#L170-L189

**Comment by veprbl** [comment]
> I get a completely different error on `aarch64-darwin` when building `curl` on [29526bc](https://github.com/NixOS/nixpkgs/commit/29526bc2eb9463515d0568c05e28b8a3bec09c91):
> 
> ```
> building '/nix/store/c5d6fgwfm7vq8qyggxgyf09j3nqxl1a7-python3-minimal-3.9.9.drv'...
> unpacking sources
> unpacking source archive /nix/store/ka6ph2kwfm5g75zxxvvg8y08zryyh9b7-Python-3.9.9.tar.xz
> source root is Python-3.9.9
> setting SOURCE_DATE_EPOCH to timestamp 1636999517 of file Python-3.9.9/Misc/NEWS
> patching sources
> applying patch /nix/store/p4qar37zn2892m83b2lkp8fw5dsjkzxy-no-ldconfig.patch
> patching file Lib/ctypes/util.py
> applying patch /nix/store/r112dk8w7zvdjipki58ch00m825li7fq-virtualenv-permissions.patch
> patching file Lib/venv/__init__.py
> Hunk #1 succeeded at 402 with fuzz 2 (offset 23 lines).
> applying patch /nix/store/dkb2rjyj7lwmvsn4zzwx85kx8r61nk9w-darwin-libutil.patch
> patching file Modules/posixmodule.c
> Hunk #1 succeeded at 7148 (offset 1268 lines).
> applying patch /nix/store/x6aiw4vay2b63slqz5byimn0alhg5b1s-darwin-tcl-tk.patch
> patching file setup.py
> Hunk #1 succeeded at 1988 (offset 7 lines).
> applying patch /nix/store/xid1jwdpp3zvg7a9zl41snb9hgvam1w3-bpo-44689-ctypes.util.find_library-now-finds-macOS-1.patch
> patching file Misc/NEWS.d/next/macOS/2021-07-20-22-27-01.bpo-44689.mmT_xH.rst
> patching file Modules/_ctypes/callproc.c
> Reversed (or previously applied) patch detected!  Assume -R? [n]
> Apply anyway? [n]
> Skipping patch.
> 2 out of 2 hunks ignored -- saving rejects to file Modules/_ctypes/callproc.c.rej
> builder for '/nix/store/c5d6fgwfm7vq8qyggxgyf09j3nqxl1a7-python3-minimal-3.9.9.drv' failed with exit code 1
> ```

That one should have been fixed in 925bbece2f8dbfb93e2d53eef7aff30ad29df1cb

**Comment by vcunat** [comment]
Builds on Hydra, too: https://hydra.nixos.org/build/164582631
